File: 125_valid_palindrome/isPalindrome.cpp
class Solution {
public:
    bool isAlphanumeric(char ch) {
        return std::isalnum(ch);
    }
    bool isPalindrome(string s) {
        int n = s.length();

        int l = 0;
        int r = n - 1;

        while (l < r) {
            if (!isAlphanumeric(s[l])) {
                l++;
                continue;
            }
            if (!isAlphanumeric(s[r])) {
                r--;
                continue;
            }

            if (isAlphanumeric(s[l]) && isAlphanumeric(s[r])) {
                if (std::tolower(s[l]) != std::tolower(s[r])) {
                    return false;
                }

                l++;
                r--;
            }
        }

        return true;
    }
};

File: 125_valid_palindrome/isPalindrome.go
// func isPalindrome(s string) bool {
// 	n := len(s)
// 	l, r := 0, n-1

// 	for l < r {
// 		lR, rR := rune(s[l]), rune(s[r])
// 		if !isAlphanumeric(lR) {
// 			l++
// 		}
// 		if !isAlphanumeric(rR) {
// 			r--
// 		}

// 		if isAlphanumeric(lR) && isAlphanumeric(rR) {
// 			if strings.ToLower(string(lR)) != strings.ToLower(string(rR)) {
// 				return false
// 			}

// 			l++
// 			r--
// 		}
// 	}

// 	return true
// }

// func isAlphanumeric(ch rune) bool {
// 	if unicode.IsLetter(ch) || unicode.IsDigit(ch) {
// 		return true
// 	}

// 	return false
// }

package validpalindrome

import "strings"

func isPalindrome(s string) bool {
    s = strings.TrimSpace(s)
    if s == "" {
        return true
    }

    s = strings.ToLower(s)
    var sb strings.Builder

    for _, c := range s {
        if c >= 'a' && c <= 'z' || c >= '0' && c <= '9' {
            sb.WriteRune(c)
        }
    }

    s = sb.String()
    l, r := 0, len(s) - 1

    for l < r {
        if s[l] != s[r] {
            return false
        }

        l++
        r--
    }

    return true
}